Add `mutatesOnly` feature the generator entries

trueadm opened the pull request:
Release notes: adds `mutatesOnly` to generator entries to better allow for DCE

This PR adds a new option that can be passed as an optional arg to `generator.deriveAbstract` called `mutatesOnly `. See https://github.com/facebook/prepack/pull/2132#pullrequestreview-130509488 for details. With it, we can better dead-code eliminate generator entries when the value entries that are flagged as mutating are dead code.

The test showing this is below:

Input:
```js
global.f = function() {
  var x = global.__abstract ? global.__abstract({}, "({})") : {};
  global.__makeSimple && __makeSimple(x);
  Object.assign({}, x);
  return 1;
};

if (global.__optimize) __optimize(f);
```

Output (with master):
```js
  var _$2 = this;

  var _$3 = _$2.Object;
  var _$4 = _$3.assign;

  var _0 = function () {
    var _$0 = _$4({}, {});

    return 1;
  };
```

Output (with the changes in this PR):
```js
  var _0 = function () {
    return 1;
  };

  var _2 = function () {
    return global.f();
  };
```

gaearon commented on src/utils/generator.js:
```diff
@@ -136,8 +137,12 @@ class TemporalBuildNodeEntry extends GeneratorEntry {
 
   visit(callbacks: VisitEntryCallbacks, containingGenerator: Generator): boolean {
+    if (this.purityCheck) {
+      this.isPure = this.purityCheck(callbacks, this.declared, this.args);
```
It's unexpected to me that the first visit mutates `this.isPure`.

trueadm replied:
It's mutated so that in the serialization phase, we know it's pure. We can't do the same purity check here, as that's only something that occurs in the visitor phase. I guess we could make another flag?

NTillmann commented on src/utils/purity.js:
```diff
@@ -0,0 +1,35 @@
+    // If the "to" value in the temporal Object.assign call is
+    // not used as reference and it is conrete/abstract, we can safely
+    // remove the Object.assign temporal altogether
+    if (to instanceof ConcreteValue || to instanceof AbstractValue) {
```
If not a `ConcreteValue` or an `AbstractValue`, what could `to` be?

trueadm replied:
It was to make Flow happy.

NTillmann commented on src/utils/purity.js:
```diff
@@ -0,0 +1,35 @@
+    // that might possibly throw at runtime.
+    // If they are, then we can proceed to checking the "to" value
+    for (let source of sources) {
+      if ((source instanceof AbstractObjectValue || source instanceof ObjectValue) && !source.isSimpleObject()) {
```
These checks could be covered with the existing `isPure` scheme.

trueadm replied:
How do you mean? We should add this to the existing scheme or it should be handled with another implementation?

NTillmann commented on src/utils/purity.js:
```diff
@@ -0,0 +1,35 @@
+    // not used as reference and it is conrete/abstract, we can safely
+    // remove the Object.assign temporal altogether
+    if (to instanceof ConcreteValue || to instanceof AbstractValue) {
+      return callbacks.canSkip(to);
```
So, this is the essence of this change? Peeking into the args and declaring an Object.assign call to be "pure" if the `to` value "can be skipped", which is when the value hasn't been visited yet, i.e. isn't subsequently used?

gaearon replied:
Yep.

trueadm commented:
@NTillmann Good points, glad I'm on the right direction with this PR then. I'll likely come back to this PR some time in the future, as I have far too many other things to be working on now. So if either you or @gaearon have any bandwidth to tidy and clean up this PR, feel free!

NTillmann commented:
Sounds good. I'll do it when I find time, but probably not this week.

trueadm commented:
I've addressed @NTillmann's feedback regarding `mutatesOnly` and changed this PR to reflect his comments. I've also fixed the Flow issues and took in the note from @cblappert in regards to `mightBeSkippable`.

NTillmann commented on src/utils/generator.js:
```diff
@@ -136,9 +137,20 @@ class TemporalBuildNodeEntry extends GeneratorEntry {
   buildNode: void | GeneratorBuildNodeFunction;
   dependencies: void | Array<Generator>;
   isPure: void | boolean;
+  mutatesOnly: void | Array<Value>;
```
In the constructor, can you add invariant(s) that ensure that `isPure` and `mutatesOnly` are mutually exclusive? Also, an invariant that the values listed in `mutatesOnly` are a subset of the `args`.

sebmarkbage commented:
Nice! This is just what I needed.
